kernel/typing.ml

```diff
@@ -152,6 +152,56 @@ let unshift_reduce sg q t =
     ( try Some (Subst.unshift q (R.snf sg t))
       with Subst.UnshiftExn -> None )
 
+exception VarSurelyOccurs
+
+(** Under [d] lambdas, checks whether term [te] *must* contain an occurence
+    of any variable that satisfies the given predicate [p],
+    *even when substituted or reduced*.
+    This check make no assumption on the rewrite system or possible substitution
+    - any definable symbol are "safe" as they may reduce to a term where no variable occur
+    - any applied meta variable (DB index > [d]) are "safe" as they may be
+      substituted and the reduce to a term where no variable occur
+    Raises VarSurelyOccurs if the term [te] *surely* contains an occurence of one
+    of the [vars].
+ *)
+let sure_occur_check sg (d:int) (p:int -> bool) : term -> unit =
+  let rec aux k = function (* k counts the number of local lambda abstractions *)
```

Comment:
Same comment here as for gather_free_vars .

Reply:
Same answer